PXC-702: GCache encryption

kamil-holubicki opened the pull request:
**Note: I intentionally haven't squashed commits yet to make it easier to review. Once approved, I will reorder commits, and squash what is needed to be squashed.**

PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Introduced EncMMap class which is intercepting MMap, providing
data at rest encryption functionality.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Introduced EncMMap class which is intercepting MMap, providing
data at rest encryption functionality.

dutow commented on galera/src/ist_proto.hpp:
```diff
@@ -668,15 +668,27 @@ namespace galera
+                            // avoid allocating huge buffers, so receive it in chunks
+                            static const ssize_t recv_buf_size = 32* 1024;
+                            unsigned char recv_buf[recv_buf_size];
+                            unsigned char* write_ptr = reinterpret_cast<unsigned char*>(ptr);
```
A static_cast should work too for a void -> char conversion

kamil-holubicki replied:
yup

dutow commented on galera/src/ist_proto.hpp:
```diff
@@ -668,15 +668,27 @@ namespace galera
                             void*   const ptr(gcache_.malloc(wsize));
-                            ssize_t const r
-                                (socket.read(gu::AsioMutableBuffer(ptr, wsize)));
+                            // avoid allocating huge buffers, so receive it in chunks
```
Why? It's faster this way, or is there a performance/stability reason behind this?

kamil-holubicki replied:
ptr is the pointer from GCache. So this is the pointer from mmap. We rely on signals to populate encryption cache pages. Unfortunately when ptr is passed to the socket (back to the syscall), signals are not notified. I didn't investigate why, but this mechanism simply doesn't work. So that's why we need this intermediate buffer (recv_buf) where we receive data from the socket and then memcpy to mmapped pointer.
Why in chunks? In general, there is no limit for the transfer size, the writeset can be huge, so we receive the data in 32kB chunks (we can increase chunk size if you think it is reasonable, but I think is fine).
Please note that this part of the code is used only during IST. "Normal" replication goes another way.

dutow commented on galera/src/replicator_smm.cpp:
```diff
@@ -207,6 +212,12 @@ galera::ReplicatorSMM::ReplicatorSMM(const struct wsrep_init_args* args)
     if (abort_cb_) gu_abort_register_cb(abort_cb_);
 #endif /* PXC */
+    // Maybe not the most elegant solution, but let's go this way
+    // as encryption configuration cannot change in runtime,
```
Shouldn't be some kind of sanity check about this? E.g. a static variable in configure_encryption or something, that it doesn't change later. (or maybe also a deconfigure_encryption that only changes the "configured" status for the sanity check, used in the destructor here?)

kamil-holubicki replied:
Implemented proposed guard + `Allocator::param_set()` method that prevents setting read only parameters.

dutow commented on galera/src/replicator_smm.cpp:
```diff
@@ -2399,6 +2410,10 @@ void galera::ReplicatorSMM::process_vote(wsrep_seqno_t const seqno_g,
 }
 
+wsrep_status_t galera::ReplicatorSMM::rotate_gcache_key() {
+    bool res = master_key_provider_.NotifyKeyRotationObserver();
```
The default observer returns true, which means this function will return WSREP_FATAL. Shouldn't the default always return false?

kamil-holubicki replied:
I think it doesn't matter because the observer is RingBuffer::RingBuffer. When it exists (and it exists always) it registers itself and returns true if encryption is disabled (rotation will cause an error) or false in case of encryption is enabled and rotation process success.
When RingBuffer is destroyed (destructor), the observer is set to the one returning `false` by default, and I think that part is wrong. In other words, for sanity, the default observer should return true (error) as it does not do any rotation. (but again: there is always RingBuffer observer registered, so only this part matters)

kamil-holubicki replied:
changed default observer to return `true` (failure in rotation)

dutow commented on galera/src/replicator_smm.cpp:
```diff
@@ -3801,3 +3816,38 @@ galera::ReplicatorSMM::abort()
+std::string
+galera::ReplicatorSMM::get_encryption_key(const std::string& keyId)
+{
+    static size_t KEY_LENGTH = 32;
```
const(expr)

kamil-holubicki replied:
done

dutow commented on galerautils/src/enc_stream_cipher.cpp:
```diff
@@ -0,0 +1,233 @@
+/* Copyright (c) 2018, 2021, Oracle and/or its affiliates.
```
Percona?

kamil-holubicki replied:
I'm not sure about this. This is the class stolen from sql/stream_cipher.cc (which is used for binlog cache encryption). I added some modifications/improvements to it, but it still originates from Oracle

dutow replied:
We have several files with both percona and oracle copyright in different lines

kamil-holubicki replied:
Added Percona copyright info.
Copyright headers added to other files as well.

dutow commented on galerautils/src/gu_asio.hpp:
```diff
@@ -482,7 +482,20 @@ namespace gu
             if (b->size() > 0)
             {
-                written += socket.write(AsioConstBuffer(b->data(), b->size()));
+                // avoid allocating huge buffers, so send it in chunks
```
This is duplicated, can't be extracted to a common function?

kamil-holubicki replied:
The other part in ist_proto.hpp is about socket.read() and is similar, but slightly different.
Or something else is duplicated? Could you elaborate, please?

dutow replied:
You are right, these are a bit different, I didn't notice that beforee.

dutow commented on galerautils/src/gu_enc_debug.cpp:
```diff
@@ -0,0 +1,42 @@
+namespace gu {
+void swrite(const char* format, ...)
+{
+#define buffer_len 8*1024
```
const

kamil-holubicki replied:
fixed

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+
+namespace gu {
+
+#define IS_LAST_PAGE(_page) (_page == pagesCnt_-1)
```
this could be a simple function

kamil-holubicki replied:
done

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+    unsigned char*   start_;
+    unsigned char*   end_;
+};
+std::map<EncMMap*, EncMMapDescriptor> encMMaps;
```
Sholudn't be functions dealing with encMMaps directly be in a class, and encMMaps could be private?

kamil-holubicki replied:
Do you mean encMMaps being a member of EncMMap class?
encMMaps is the container of all existing EncMMap objects. The new EncMMap object is added when mmap is created, deleted when mmap is destroyed and accessed/searched in the `signal_handler` function which is called by the OS. When proper EncMMap object is called, signal handling is redirected to this object.
encMMaps and functions dealing with this container sit in between of actual EncMMap object and the signal handler, so there is no logical reason of this stuff being owned by EncMMap class.

dutow replied:
I meant that the global encmap functions + encmmap + the atomic flag and other related things could be a single instance class, with enmmap and the flag being private.

kamil-holubicki replied:
done. EncMMapsRepository introduced.

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+    while (encMmapsLock.test_and_set(std::memory_order_acquire)) {
+        std::this_thread::yield();
+    }
+    encMMaps[mmap] = {ptr, ptr+size};
```
emplace

kamil-holubicki replied:
done

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+}
+
+static EncMMap* getEncMMap(unsigned char* ptr) {
+    for (auto m : encMMaps) {
```
missing encMMapsLock use

kamil-holubicki replied:
It is not missing. It is done this way to not lock conflicting signals. Please check the logic and the detailed explanation in `signal_handler` function.

dutow replied:
Yes, I see now. Then I would do something here to make the API cleaner - because right now, other functions do locking, so this is different.

Maybe this function could be tryGetEncMMap instead, and return null if it can't get the lock?

kamil-holubicki replied:
done

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+PMemoryManagerPool memoryManagerPool(MANAGERS_POOL_SIZE);
+
+// EncMMap objects repository
+static std::atomic_flag encMmapsLock = ATOMIC_FLAG_INIT;
```
have you measured this vs a std::mutex?

kamil-holubicki replied:
We cannot use mutex here as we are in signal handler. std::atomic_flag is guaranteed to be lock-free.

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+        try { unmap(); } catch (Exception& e) { log_error << e.what(); }
+    }
+
+    delEncMMap(this);
```
Sholdn't this happen before unmap? Now it's possible that we receive a signal when it was unmapped, but not deleted yet from the map.

kamil-holubicki replied:
`unmap()` call still needs `this` object to be present in encMMaps container because it will potentially do syncing, which will potentially trigger signals. 
On the other hand, if we receive signal after `munmap` it means that the application is doing something really wrong, because it is destroying MMap object and at the same time accessing it...
I think we can improve it in the way that removing from encMMaps container will happen after `sync()`, but before actual `munmap`.

kamil-holubicki replied:
Removal from encMMaps map moved after sync and before actual unmap. It is still up to the application to not use MMapEnc object while deletion/unmap, because it will cause an unhandled segfault anyway.

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+}
+
+bool EncMMap::lock() {
+    return !lock_.test_and_set(std::memory_order_acquire);
```
Same question: is this significantly faster than a mutex?

kamil-holubicki replied:
We are in the signal handler, so no mutex is allowed.

dutow commented on galerautils/src/gu_enc_pmemory_manager.cpp:
```diff
@@ -0,0 +1,175 @@
+      ptr2ull(base_), ptr2ull(base_) + size_);
+
+    if (freePages_.size() != allocPagesCnt_) {
+        S_DEBUG_A("Some pages still allocated. Free pages cnt: %d\n", freePages_.size());
```
Shouldn't be this a higher level message? So it's easier to notice it in MTR for example if it ever happens.

kamil-holubicki replied:
S_DEBUG macros are intended to be used from the signal handler context. We can't use fprintf/existing server logging system, because we need to be lock-free. So unistd.h write() function needs to be used.
By "higher level" do you mean formatting the message like server's [ERROR]/[WARNING], to be automatically catchable by MTR framework?

dutow replied:
Yes, I meant the error/warning formatting.

kamil-holubicki replied:
Implemented.

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+    unsigned char *addr = static_cast<unsigned char*>(info->si_addr);
+    S_DEBUG("addr: x%llX\n", ptr2ull(addr));
+    EncMMap*  encmmap = getEncMMap(addr);
+    encMmapsLock.clear(std::memory_order_release);
```
Is it safe to clear this here? In theory, other threads could delete encmmap after this, since the signal handler isn't holding a shared_ptr.

kamil-holubicki replied:
We are clearing it here to allow other threads which use other EncMMap objects to work in parallel. So this only protects encMMaps container from new objects additions/deletions which can happen in parallel.
EncMMap object related to the current signal cannot be deleted while the signal is active because that would mean that the application is destroying EncMMap and at the same time is using it.

dutow commented on galerautils/src/gu_enc_mmap.cpp:
```diff
@@ -0,0 +1,628 @@
+    // 2. (more probable but still not so often) Two signal handlers are called
+    //    simultaneously (2 threads). In such a case the 2nd one will retry
+    //    in a while.
+    if (encMmapsLock.test_and_set(std::memory_order_acquire)) {
```
We have to use test_and_set because we can't use mutexes in signal handlers?

In that case, it would be useful to add a RAII helper to this, similar to scoped_lock.

kamil-holubicki replied:
Yes, this is the reason.
As for RAII helper, could you give the example how it would work?
scoped_lock locks in its constructor. There is no possibility to NOT lock. It waits until the lock is possible, then goes on, so once the object is created, it is locked.
Here we are locking or not (something like try-lock). 
How to RAII it? Do you mean something like:
```
std::unique_lock<std::mutex> lock(_mutex, std::try_to_lock);
if(!lock.owns_lock()){
    // mutex wasn't locked. Handle it.
}
```
Is it worth it? Please check `addEncMMap` and `delEncMMap`. I'm not sure it will make it any simpler/safer

dutow replied:
Yes, it would make sense only for addEncMMap and delEncMMap. I know those are only small functions, so maybe its not worth it.

kamil-holubicki replied:
This part has been reworked.

dutow commented on gcache/src/gcache_rb_store.cpp:
```diff
@@ -704,14 +786,116 @@ namespace gcache
+               be ideal place, but again, we don't have access to it when Galera
+               starts, so another solution is keeping it in dedicated file.
+               But having yet another file on server side if we have this preamble
+               does not seem to be good idea. */
```
Isn't the preamble part of the actual gcache? It's in the same memory block, only separated by an offset.

kamil-holubicki replied:
Yes, RingBuffer file already starts with the preamble.
I'm not sure I understand the question. Could you elaborate?

dutow replied:
Sorry, I forgot to edit this question after reading further, the issue here is with the comment, not the code: the master key isn't stored in the preamble, only its name.

kamil-holubicki replied:
comment ammended

dutow reviewed (commented):
Some initial comments / questions, I'll add more later.

dutow commented on gcache/src/gcache_page_store.hpp:
```diff
@@ -24,7 +24,10 @@ namespace gcache
-                   size_t             keep_page);
+                   size_t             keep_page,
+                   bool               encrypt,
+                   size_t             encryptCachePageSize,
```
This is just a formatting issue, but it seems like galera is using underscored_names for variables, while this patch uses camelCase.

kamil-holubicki replied:
Changed everything to use underscored_names

dutow commented on galerautils/src/gu_enc_utils.cpp:
```diff
@@ -0,0 +1,148 @@
+
+    const unsigned char* keyPtr = reinterpret_cast<const unsigned char*>(key.c_str());
+    const unsigned char* keyToBeDecryptedPtr = reinterpret_cast<const unsigned char*>(keyToBeDecrypted.c_str());
+    char resultBuf[gu::Aes_ctr::FILE_KEY_LENGTH];
```
Since this is an array allocated on the stack, I wouldn't leave this length check to just an assertion - the decryption itself probably takes way more time than a simple if check to ensure we have a big enough buffer.

(the function above has the same issue)

kamil-holubicki replied:
Changed as suggested

dutow commented on gcache/src/gcache_rb_store.cpp:
```diff
@@ -704,14 +786,116 @@ namespace gcache
+                    masterKeyId_ = 1;
+
+                    mkName = gu::CreateMasterKeyName(masterKeyUuid_, masterKeyId_);
+                    masterKeyProvider_.CreateKey(mkName);
```
Sholudn't this error out based on the return value of CreateKey?

kamil-holubicki replied:
this part has been reworked

dutow commented on gcache/src/gcache_rb_store.hpp:
```diff
@@ -204,9 +218,15 @@ namespace gcache
 
         void          write_preamble(bool synced);
-        void          open_preamble(bool recover);
+        void          open_preamble(bool const recover);
```
Why connst, this isn't a reference?

kamil-holubicki replied:
fixed

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Existing Galera unit tests resurrected.
1. Compilation fixed
2. Tests failures related to GCache encryption modifications fixed
(initialization of Galera Utils - CRC utils needed by RingBuffer).
Note that tests pass only with Galera encryption disabled (these are old
tests adjusted to the new code). For them to work with encryption
enabled additional work is needed.
3. There are 3 tests failing inherited from trunk:
test_out_queue_limit
test_certification_trx_v3
Provider version string '4.11(XXXX)' not expected '26.4.11(rXXXX)'
It has been reported under PXC-4020

dutow commented on galerautils/src/gu_enc_utils.cpp:
```diff
@@ -0,0 +1,163 @@
+    static const std::string MASTER_KEY_SEQNO_SEPARATOR = "-";
+    static const std::string MASTER_KEY_ID_SEPARATOR = "@";
+    std::ostringstream os;
+    os << uuid << MASTER_KEY_ID_SEPARATOR << const_uuid;
```
This is just a formatting question, but why does this function mix string construction with << and +? It would be cleaner to use the stringstream for everything, and then the static const could be plain const char*s.

kamil-holubicki replied:
Done

dutow commented on galerautils/src/gu_enc_utils.hpp:
```diff
@@ -0,0 +1,53 @@
+#define ptr2ull(ptr) ((unsigned long long)ptr)
+
+std::string encode64(const std::string& binary);
+std::string decode64(const std::string& base64);
```
Also just a stylistic comment, but this pull request uses multiple styles for functions/classes, sometimes even within a single file.

Like here, we have two function stating with small letters, and then 4 below using CamelCase.

kamil-holubicki replied:
All method names were changed to `underscore_convention()`.
private and public methods follow the same pattern.
Only constructor/destructor use CammelCase

It looks like this is the Galera convention, so let's stick to it, indeed.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Fixed issues found during testing:
PXC-4025, PXC-4027, PXC-4031, PXC-4035, PXC-4036, PXC-4044, PXC-4045

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Refactoring (mainly adjust methods/variable names to be compliant with
the coding standard)

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Force at least 2 pages in encryption cache.
This is because how memcpy copies memory in
16-bytes memory chunks. If we are on virtual pages boundary, both pages
have to be mapped. If we have only one cache page it is not possible.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

wsrep-api submodule pointer updated
